Docker compose support

schuellerf opened the pull request:
Thanks for this project, this really works great!

For my usecase I needed support to generate short and more handy DNS names for compose-services.

like "my_app.docker"

The implementation is rather rude, please give me feedback if it works for you and which part I should refactor to be nicer to read. Especially that _all_ "listen_addresses" are handed over to each SystemdResolvedConnector is not really nice...

I also added a service file for debian which probably should be split into a separate pullrequest and leads to the idea to implement proper DEB packaging...

schuellerf committed:
Document docker-compose behavior

flaktack reviewed (commented):
Thanks for the contributions! Do you have an example docker-compose file which I could use to try out, better understand the new features?

schuellerf commented on README.md:
```diff
@@ -74,6 +74,14 @@ If configured correctly then `resolvectl status` should show the configured link
         DNS Domain: ~docker
     ...
 
+### Docker compose support
+
+For containers started with docker-compose the service tries to generate a short domain name: `<compose-service-name>.<default_domain>`
+
+This would result in a name like `my_app.docker`
+
+Once you would use two compose files with the same service name or start to scale out the service then the short form `<compose-service-name>.<default_domain>` will be removed and only the unique names as described above (e.g. `<container_name>.<container_network>.<default_domain>`) will be available.
```
I'd also like this to work "out of the box" as it's just an additional name to be available (unless you start multiple compose files with the same service name...) but guarding this with some environment variable is possible for sure

schuellerf commented:
> Thanks for the contributions! Do you have an example docker-compose file which I could use to try out, better understand the new features?

A file `docker-compose.yml` with
```
version: "2.1"
services:
  webserver:
    image: nginx
  broker:
    image: redis
```
and starting this in foreground with `docker-compose up` (in the same directory as `docker-compose.yml`) should be sufficient
then you can `ping webserver.docker` or `ping broker.docker`

schuellerf committed:
Support multiple bridges for docker-compose

schuellerf commented:
You can see the "behavior" when you have the docker-compose running (from my last message) and then do
`docker-compose scale broker=2` (in a second terminal or you did `docker-compose up -d` before...)

Then "broker.docker" would not be unique anymore and gets removed. I think users who "scale" know that "broker.docker" won't work and the won't be using it anyway...

as an example, I reformatted the logs of systemd-resolved-docker to make this more visible what happens:

docker-compose up:
```
Mai 12 16:17:54 mylaptop systemd-resolved-docker[23979]: Refreshed - 2 items
  172.18.0.3/b9369ca8826f.docker  broker.docker     tmpdp08ldplvj_broker_1.tmpdp08ldplvj_default.docker
  172.18.0.2/783ce4b8fb7e.docker  webserver.docker  tmpdp08ldplvj_webserver_1.tmpdp08ldplvj_default.docker
```
docker-compose scale broker=2
```
Mai 12 16:18:27 mylaptop systemd-resolved-docker[23979]: Refreshed - 3 items
  172.18.0.4/4d26c16c6b91.docker  tmpdp08ldplvj_broker_2.tmpdp08ldplvj_default.docker
  172.18.0.3/b9369ca8826f.docker  tmpdp08ldplvj_broker_1.tmpdp08ldplvj_default.docker
  172.18.0.2/783ce4b8fb7e.docker  webserver.docker  tmpdp08ldplvj_webserver_1.tmpdp08ldplvj_default.docker
```

(I now removed the debian systemd file from this branch)

schuellerf committed:
Fix racecondition when "scaling down"

When "scaling down to 1 instance" the short-name (like my_app.docker) would not be
re-added as the "container #2" seems to be still there (shutting down) when
evaluating if the name is unique.

flaktack commented:
Thanks for the changes, after some thought / testing there is no need for the environment variable guard. :-)

One last question: why do multiple interfaces / bridges need to be handled? If the default bridge is not used, then docker takes care of proxying the DNS requests, and so the DNS proxy implemented here is not needed.

The following compose file works without listening on multiple interfaces:
```
version: "2.1"
services:
  webserver:
    image: nginx
    networks:
      - pytest
  broker:
    image: redis
    networks:
      - pytest

networks:
  pytest:
    driver: bridge
    enable_ipv6: false
    ipam:
      driver: default
      config:
        - subnet: 172.16.238.0/24
          gateway: 172.16.238.1
```

schuellerf commented:
> One last question: why do multiple interfaces / bridges need to be handled? If the default bridge is not used, then docker takes care of proxying the DNS requests, and so the DNS proxy implemented here is not needed.

I had problems when leaving "systemd-resolved-docker" running and starting stopping different containers that the request just did not arrive (at the systemd-resolved-docker). I'll try to reproduce and write down the steps - This is the area why I think the code can be improved/stripped down...

flaktack commented:
If you can reproduce it could you also check if the DNS server is correctly configured for the `docker0` interface using `resolvectl status`. It  could be that `systemd-resolved` "forgets" the configuration.

schuellerf committed:
Add some test files for docker-compose

schuellerf commented:
I think my problem is really more related to docker-compose creating new docker-networks, where `resolvectl status` looks like

```
Link 65 (br-f6139bfdac9f)
      Current Scopes: none
DefaultRoute setting: no  
       LLMNR setting: yes 
MulticastDNS setting: no  
  DNSOverTLS setting: no  
      DNSSEC setting: no  
    DNSSEC supported: no  

Link 7 (docker0)
      Current Scopes: none      
DefaultRoute setting: no        
       LLMNR setting: yes       
MulticastDNS setting: no        
  DNSOverTLS setting: no        
      DNSSEC setting: no        
    DNSSEC supported: no        
         DNS Servers: 172.17.0.1
          DNS Domain: ~docker   
```
vs this after restarting systemd-resolved-docker
```
Link 65 (br-f6139bfdac9f)
      Current Scopes: DNS         
DefaultRoute setting: no          
       LLMNR setting: yes         
MulticastDNS setting: no          
  DNSOverTLS setting: no          
      DNSSEC setting: no          
    DNSSEC supported: no          
         DNS Servers: 172.16.238.1
                      172.17.0.1  
          DNS Domain: ~docker     

Link 7 (docker0)
      Current Scopes: none        
DefaultRoute setting: no          
       LLMNR setting: yes         
MulticastDNS setting: no          
  DNSOverTLS setting: no          
      DNSSEC setting: no          
    DNSSEC supported: no          
         DNS Servers: 172.16.238.1
                      172.17.0.1  
          DNS Domain: ~docker     
```

I added some test docker-compose files in my repo and will probably have to extend the `DockerWatcher` to also handle 'create' and 'destroy' events of `'Type' == 'network'`

Test files: https://github.com/schuellerf/systemd-resolved-docker/tree/master/test

schuellerf committed:
Implement dynamic allocation of DNS servers and Systemd registrations

schuellerf commented:
Hi!
I kind of reworked the code so the registration at systemd and the creation of the DNS servers is now dynamic.

Please test if this works for you.

Finally I'd update the documentation and I'm not really sure if overriding the DOCKER_INTERFACE and LISTEN_ADDRESS is really beneficial as it's now dynamic anyway.
If you are fine with the code I'd rather document that those options (DOCKER_INTERFACE and LISTEN_ADDRESS) are only for testing and not useful in a real scenario.

schuellerf committed:
Add initial update of hostname list

flaktack commented:
I've finally had time to take a look -- using the above `test/sample` test I found the following when the DNS could not be resolved: Is it possible that when the DNS resolving failed for you there were no containers running in the default network? (The above `sample` docker-compose has its own network, and so the default network stays empty when it is used).

The `docker0` interfaces is only `UP` when there are containers actually using it (`ip link show docker0`).
```
6: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN mode DEFAULT group default 
    link/ether 02:42:6a:1b:6d:4f brd ff:ff:ff:ff:ff:ff
```

Even if the DNS server is configured, it is not used while the link is down (`resolvectl status docker0`):
```
Link 6 (docker0)
Current Scopes: none
     Protocols: -DefaultRoute +LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
   DNS Servers: 172.17.0.1
    DNS Domain: ~docker
```

If a container is started (`docker run --rm -it alpine`) than both the status changes to `UP` and the _Current Scopes_ is updated:
```
6: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default 
    link/ether 02:42:6a:1b:6d:4f brd ff:ff:ff:ff:ff:ff
```
```
Link 6 (docker0)
    Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
         Protocols: -DefaultRoute +LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
Current DNS Server: 172.17.0.1
       DNS Servers: 172.17.0.1
        DNS Domain: ~docker
```

flaktack commented:
In 9a686219 I've added a GitHub Action to build a `*.deb` file -- could you check if it works? The created file can be downloaded from the workflow artifacts (https://github.com/flaktack/systemd-resolved-docker/actions/workflows/build-packages.yml). It currently uses the Fedora `*.service` file, but that can be replaced if needed with a debian/ubunti specific one.

flaktack commented:
In the [feature/split-dns](https://github.com/flaktack/systemd-resolved-docker/tree/feature/split-dns) branch I've added an alternative approach: instead of registering the DNS server for every docker interface, a dummy interface is created which seems like a simpler approach.

Does using the dummy interface work for you?

I've also changed the generated `docker-compose` names to be `<service>.<project>` (and `<container number>.<service>.<project>`) format. If a service is scaled, then there will be multiple responses to the first query. This matches expectations when scaling and avoids the need for handling duplicates.

With the `sample` project this results in the following names (`docker-compose up -d --scale webserver=2`):
```
172.16.238.2/9db5e667e6f3.docker,webserver.sample.docker,1.webserver.sample.docker,sample_webserver_1.sample_pytest.docker
172.16.238.3/4f225c0db0a9.docker,webserver.sample.docker,2.webserver.sample.docker,sample_webserver_2.sample_pytest.docker
172.16.238.4/903366b695bc.docker,broker.sample.docker,1.broker.sample.docker,sample_broker_1.sample_pytest.docker
```

schuellerf commented:
The package doesn't work in ubuntu 20.04.4 (at least three problems)
 * Dependencies are "python3-docker python3-dnslib python3-dbus python3-pyroute2 python3-systemd" (always _with_ python**3** in this distro version)
 * the systemd-resolved-docker.service references  /usr/bin/systemd-resolved-docker but the package installs /usr/local/bin/systemd-resolved-docker
 * when executing python does not find the package "importlib.metadata.PackageNotFoundError: systemd-resolved-docker"
I'm not experienced in packaging how to fix that properly :-/

flaktack commented:
I've updated the built DEB package and added a job to test that it installs and runs correctly on Ubuntu 20.04 and 22.04:
https://github.com/flaktack/systemd-resolved-docker/actions/runs/2564696404

schuellerf commented:
The package now works perfect for me!

Only one error appears (which does not affect functionality)
`ERROR:dbus.connection:Unable to set arguments (9, [[<AddressFamily.AF_INET: 2>, b'\x7f\x00\x00\x99', 53, '']]) according to signature None: <class 'TypeError'>: an integer is required (got type bytes)`
that seems to be easy to fix...
And your branch `feature/split-dns` does not implement my very handy shortcut to also register "<service>.<DEFAULT_DOMAIN>" additionally to "<service>.<project>.<DEFAULT_DOMAIN>" (if still unique)

I think I'd rather close this pull-request and re-implement the "DNS shortcut names" on your current `feature/split-dns` branch. Does that make sense?

schuellerf commented:
btw. @flaktack I also tested the DEB package (your branch) on my server where there is only one docker compose with "network_mode: host" which then resolved to "127.0.0.1" which is correct and really perfect :-D

127.0.0.1/6c5a5cf34fd5.docker,tower.docker,timescaledb.iobroker.docker,1.timescaledb.iobroker.docker,iobroker_timescaledb_1.host.docker,iobroker.host.docker

This also includes (as some sideeffect?) the real hostname "tower" as "tower.docker" and some "iobroker.host.docker" which is fine I guess and does not hurt :+1:

flaktack commented:
Thanks for testing!

> Only one error appears (which does not affect functionality)

I've created #4 to track the error message, which is due to an older version of systemd.

> And your branch feature/split-dns does not implement my very handy shortcut to also register `<service>.<DEFAULT_DOMAIN>`

I've restored the functionality.
